### DISCUSSION TITLE
Best practices for logging in setup.py

### DISCUSSION THREAD

**Discussion by mwtoews** [pypa/setuptools#3710] [General]
With dynamic aspects of setup.py it is useful to see messages on (e.g.) discovery of libraries to compile modules. What is the best practice to show these messages?

I've been looking at [setuptools/tests/test_logging.py](https://github.com/pypa/setuptools/blob/v65.6.2/setuptools/tests/test_logging.py) for inspiration, but it only tests the internal logging. Could this logger be extended for other purposes? The standard [logging](https://docs.python.org/3/library/logging.html) module (now used by setuptools) is an obvious way, but it's not clear how it should be used.

### Attempt 1
This attempt only shows "BEFORE setup()" from log threshold WARNING and higher before `setup()` and then for "MESSAGE TWO" either DEBUG or higher with `--verbose` or INFO or higher without.
```python
import logging
from setuptools import setup

log = logging.getLogger()

log.debug("BEFORE setup() DEBUG")
log.info("BEFORE setup() INFO")
log.warning("BEFORE setup() WARNING")

setup(name="test_logging", version="0.1")

log.debug("AFTER setup() DEBUG")
log.info("AFTER setup() INFO")
log.warning("AFTER setup() WARNING")
```
Here it is with `--verbose`:
```
$ python setup.py --verbose sdist
BEFORE setup() WARNING
... running sdist ...
AFTER setup() DEBUG
AFTER setup() INFO
AFTER setup() WARNING
```
### Attempt 2
This uses non-public API and is a hack.
```python
import logging
from setuptools import setup
from setuptools import logging as setuptools_logging

setuptools_logging.configure()
log = logging.getLogger()

log.debug("BEFORE setup() DEBUG")
log.info("BEFORE setup() INFO")
log.warning("BEFORE setup() WARNING")

setup(name="test_logging", version="0.2")

log.debug("AFTER setup() DEBUG")
log.info("AFTER setup() INFO")
log.warning("AFTER setup() WARNING")
```
and now without `--verbose`:
```
$ python setup.py sdist
BEFORE setup() DEBUG
BEFORE setup() INFO
BEFORE setup() WARNING
running sdist
...
AFTER setup() INFO
AFTER setup() WARNING
```
### Analysis and expected behaviour
With both of the examples, the logger behaves as expected after `setup()`, as it has set the appropriate log level based on the presence (or absence) of a verbose flag. However, these messages are used to show messages gathering data before `setup()`. Neither of the above two attempts show messages for the appropriate logger level. Is there a better way?

And as a bonus question, why does the placement order of `--verbose` matter? E.g. `python setup.py sdist --verbose` ignores the verbose flag.

**Comment by webknjaz** [comment]
I haven't seen anybody caring about such logging, except for an occasional exception message being longer.

But FWIW abusing `setup.py` as a command wrapper (i.e. calling it directly) is highly discouraged and only the use via the PEP 517 interface is supported, any other misuses are deprecated.
https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html

**Reply by mwtoews** [reply]
Yes, I'm aware calling setup.py is deprecated, just providing simple demos. Similar outputs can be obtained via `python -m build --sdist`

**Reply by webknjaz** [reply]
Well, I haven't seen anything more involved than this https://github.com/pyca/cryptography/blob/d7d6b19/setup.py#L80-L91.
